### PR TITLE
Tokio bump

### PR DESCRIPTION
**Pull request by CollinValley**
This is a huge change, it needs some refactoring, particularly in the harness area.
Will probably layer on/rework, take a look at the harness code in particular, I think that is the most ugly of all of the reasons. But yeah, the bump actually worked!

Drew in particular the harness/test code could use some magic. The biggest gotcha right now is the PacketStreams, it they use a timer, need to be built in a function with the macro Tokio::main over it. That isn't amazing and maybe we can figure out how to work around that.

### CODE REVIEW

**Comment by CollinValley** [comment]
Wrote in the test harness changes, and standardized all of them. It was quite a lot of work changing the setup and teardown of every test. But they all pass now, and we even solved a multithreaded runtime bug along the way

Next I need to resolve the rest cases (2) that are failing in Graphgen right now. If the travis checks pass something isn't working.  To do that I need to sit down for awhile with graphgen, which is something I've needed to do anyways. 

I'll try to have the graphgen tests fixed in the next couple days and then we will be ready for merge pending CR!

File: examples/dns-interceptor/src/pipeline.rs
Comment (ddowl) [inline examples/dns-interceptor/src/pipeline.rs]: Construction via `map/fold`? _#unnecessarilyfunctional_

File: route-rs-runtime/src/link/composite/drop_link.rs
Comment (ddowl) [inline route-rs-runtime/src/link/composite/drop_link.rs]: Can we annotate `finishes_with_wait` with multiple macros so we don't need to dispatch to an extra function?

File: route-rs-runtime/src/link/mod.rs
Comment (ddowl) [inline route-rs-runtime/src/link/mod.rs:30]: Can you describe why `PacketStream`s are `Unpin`? Is this just explicitly stating that packets may be moved/copied across threads/cores? I thought [most types implement `Unpin` by default](https://doc.rust-lang.org/std/marker/trait.Unpin.html), but a generic type needs to be marked?

Comment (ddowl) [reply route-rs-runtime/src/link/mod.rs:30]: Maybe without `Unpin` we would need to enforce that every packet is `Pin`ned, and that's not a restriction we want to make? https://rust-lang.github.io/async-book/04_pinning/01_chapter.html#how-to-use-pinning

File: route-rs-runtime/src/link/primitive/classify_link.rs
Comment (ddowl) [inline route-rs-runtime/src/link/primitive/classify_link.rs:169]: Can we `#[derive(Unpin)]`?

File: route-rs-runtime/src/link/primitive/fork_link.rs
Comment (ddowl) [inline route-rs-runtime/src/link/primitive/fork_link.rs:143]: Why declare `PacketStream`s `Unpin` if packets are `Pin`?

Comment (ddowl) [reply route-rs-runtime/src/link/primitive/fork_link.rs:143]: I'm not asking to be snarky, I genuinely don't know.

File: route-rs-runtime/src/link/primitive/input_channel_link.rs
Comment (ddowl) [inline route-rs-runtime/src/link/primitive/input_channel_link.rs:53]: Can we `#[derive(Unpin)]`?

File: route-rs-runtime/src/link/primitive/join_link.rs
Comment (ddowl) [inline route-rs-runtime/src/link/primitive/join_link.rs:114]: Can we `#[derive(Unpin)]`?

File: route-rs-runtime/src/link/primitive/join_link.rs
Comment (ddowl) [inline route-rs-runtime/src/link/primitive/join_link.rs:214]: Comment: I like that `Future`s and `Stream`s have different poll fns `poll / poll_next` to help distinguish them.

File: route-rs-runtime/src/link/primitive/input_channel_link.rs
Comment (ddowl) [inline route-rs-runtime/src/link/primitive/input_channel_link.rs:58]: Not sure why `poll_next` must take `Pin<&mut Self>` as the first argument, since the [documentation](https://docs.rs/futures/0.2.0/futures/stream/trait.Stream.html#tymethod.poll_next) says the type bound is just `&mut Self`. However, `poll` [_is_ restricted](https://doc.rust-lang.org/std/future/trait.Future.html#tymethod.poll) to `Pin<&mut Self>`.

File: route-rs-runtime/src/link/primitive/join_link.rs
Comment (ddowl) [inline route-rs-runtime/src/link/primitive/join_link.rs]: I'd be happy to look into this later.

File: route-rs-runtime/src/link/primitive/join_link.rs
Comment (ddowl) [inline route-rs-runtime/src/link/primitive/join_link.rs]: Same as above, can we annotate a function with multiple macros?

File: route-rs-runtime/src/link/primitive/process_link.rs
Comment (ddowl) [inline route-rs-runtime/src/link/primitive/process_link.rs:99]: Lol see ya `Err` 👋

File: route-rs-runtime/src/link/utils/task_park.rs
Comment (ddowl) [inline route-rs-runtime/src/link/utils/task_park.rs]: Maybe rename to `park_and_wake` to fit new API?

File: route-rs-runtime/src/utils/test/harness.rs
Comment (ddowl) [inline route-rs-runtime/src/utils/test/harness.rs:49]: 😞

File: route-rs-runtime/src/utils/test/harness.rs
Comment (ddowl) [inline route-rs-runtime/src/utils/test/harness.rs]: Personally I'd prefer modifying `run_link` rather than leaving an old copy around. We can revisit the ugliness later if we need to.

File: route-rs-runtime/src/utils/test/harness.rs
Comment (ddowl) [inline route-rs-runtime/src/utils/test/harness.rs:34]: Initializing tests in the async runtime context is fine for now. It might be interesting to see if we could get a `#[with_tokio_runtime]` macro working so we could annotate `async fn`s. That would be a nice easy win to reduce boilerplate and 1 level of indentation.

File: route-rs-runtime/src/link/primitive/classify_link.rs
Comment (ddowl) [inline route-rs-runtime/src/link/primitive/classify_link.rs:211]: It seems like each link implements the "eggressor check" in a slightly different way. Is this due to actual different contexts among links, or are they common enough that we can abstract them into the `task_park` module?

File: examples/minimal-static-router/src/classifiers.rs
Comment (ddowl) [inline examples/minimal-static-router/src/classifiers.rs:147]: We may want to encapsulate this pattern into the test harness later

**Review by ddowl — CHANGES_REQUESTED** [review]
Overall, seems like a pretty straightforward mechanical change with a few unforeseen minor bugs. We can either address my comments or tag them for later work so we can push the version bump through. Good job @CollinValley !

File: route-rs-runtime/src/link/composite/drop_link.rs
Comment (CollinValley) [reply route-rs-runtime/src/link/composite/drop_link.rs]: This is changed in the next commit where I rework the testing harness.  I just reworked run_link and such. See the test harness commit. I think that resolves this.

Comment (ddowl) [reply route-rs-runtime/src/link/composite/drop_link.rs]: Oh yes I believe you're right. I was leaving comments as I read through each commit (they're a little more bite-sized, focused), so we can resolve them if you already addressed the comment.

File: route-rs-runtime/src/link/mod.rs
Comment (CollinValley) [reply route-rs-runtime/src/link/mod.rs:30]: This isn't that the packets need to be Unpin, it is the object that provides the Stream must be unpin. Unpin means we can freely pin and then unpin the object. Structs wont implement unpin by default because they may contain pointers and such that could move. They could provide a custom implementation if that were the case, but since the ingressor and egressor own all their components, it isn't an issue to just add the trait to get the included 'provided' methods. It is because when we call poll_next on that stream object, we need to pin it in place so the futures work correctly.  AFAIK, this is because the runtime needs to know that the object doesn't move out from underneath it when it is trying to reference the future. The poll_next function can't move on the runtime otherwise its calls would become invalidated it the function pointer they stored no longer pointed at the poll function cause the object moved in memory.  Something like that, I think the withoutboats talk may be more illuminating.  But in essence, we need our objects that implement packetstream to impl unpin so that I can shove them into a Pin object to call their poll_next() function.  Checkout the function sig of poll_next to see why.

Comment (CollinValley) [reply route-rs-runtime/src/link/mod.rs:30]: I think unpin is kind of a bad name, but I'm not sure if I can think of anything better.

File: route-rs-runtime/src/link/primitive/classify_link.rs
Comment (CollinValley) [reply route-rs-runtime/src/link/primitive/classify_link.rs:169]: Hmm...maybe.

Comment (CollinValley) [reply route-rs-runtime/src/link/primitive/classify_link.rs:169]: Looks like no, that macro doesn't exist...yet?

File: route-rs-runtime/src/link/mod.rs
Comment (CollinValley) [reply route-rs-runtime/src/link/mod.rs:30]: Okay I think this will be informative: Take a look at ClassifyIngessor's Future impl
`impl<'a, C: Classifier> Future for ClassifyIngressor<'a, C> {
    type Output = ();

    /// Same logic as QueueEgressor, except if any of the channels are full we
    /// await that channel to clear before processing a new packet. This is somewhat
    /// inefficient, but seems acceptable for now since we want to yield compute to
    /// that egressor, as there is a backup in its queue.
    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
        let ingressor = Pin::into_inner(self);
        loop {
            for (port, to_egressor) in ingressor.to_egressors.iter().enumerate() {
                if to_egressor.is_full() {
                    park_and_notify(&ingressor.task_parks[port], cx.waker().clone());
                    return Poll::Pending;
                }
            }

            //TODO: Standardize in_stream, input_stream, and stream to one name
            let packet_option: Option<C::Packet> =
                ready!(Pin::new(&mut ingressor.input_stream).poll_next(cx));

            match packet_option {
                None => {
                    for to_egressor in ingressor.to_egressors.iter() {
                        to_egressor
                            .try_send(None)
                            .expect("ClassifyIngressor::Drop: try_send to_egressor shouldn't fail");
                    }
                    for task_park in ingressor.task_parks.iter() {
                        die_and_notify(&task_park);
                    }
                    return Poll::Ready(());
                }
                Some(packet) => {
                    let class = ingressor.classifier.classify(&packet);
                    let port = (ingressor.dispatcher)(class);
                    if port >= ingressor.to_egressors.len() {
                        panic!("Tried to access invalid port: {}", port);
                    }
                    if let Err(err) = ingressor.to_egressors[port].try_send(Some(packet)) {
                        panic!(
                            "Error in to_egressors[{}] sender, have nowhere to put packet: {:?}",
                            port, err
                        );
                    }
                    unpark_and_notify(&ingressor.task_parks[port]);
                }
            }
        }
    }
}`

The first interesting line: `        let ingressor = Pin::into_inner(self);`
This ingressor, when the runtime calls it via Poll(..), has a Pinned version of the ingressor for self.  So to actually get at the internal values, we need to unpin it, so then we can refer to our methods and member variables as usual. If this were not an unpin struct, ie it was self referential, it wouldn't work. If it is self-referential or something like that, once you Pin it in place, you say "This can't be moved without invalidating something", which in the this example would be the internal pointer to itself.  Prior to Pin I don't think being self referential was something the rust lang supported. 

The second line: `ready!(Pin::new(&mut ingressor.input_stream).poll_next(cx));` Is the same thing but in reverse. Before we can call poll_next on our Stream object, we need to tell the computer to pin the Stream in place, so to satisfy the requirements of poll_next and poll, where self is `self: Pin<&mut Self>`

File: route-rs-runtime/src/link/primitive/fork_link.rs
Comment (CollinValley) [reply route-rs-runtime/src/link/primitive/fork_link.rs:143]: Packets have no Pin requirement. In your example, ForkIngressor the object must be Pinned before one may call poll on it. 
For example

```
let cx = Context; //We got this variable from somewhere, not important for this example.  
let forkingressor = ForkIngressor::new() //lets pretend this works
forkingressor.poll(cx); //Compiler error: method not found
Pin::new(&mut forkingressor).poll(cx);  //Happy rustc
```

File: route-rs-runtime/src/link/primitive/input_channel_link.rs
Comment (CollinValley) [reply route-rs-runtime/src/link/primitive/input_channel_link.rs:53]: :bigsad: no

File: route-rs-runtime/src/link/primitive/input_channel_link.rs
Comment (CollinValley) [reply route-rs-runtime/src/link/primitive/input_channel_link.rs:58]: You've been smekledorfed by old documentation https://docs.rs/futures/0.3.1/futures/stream/trait.Stream.html

File: route-rs-runtime/src/link/primitive/join_link.rs
Comment (CollinValley) [reply route-rs-runtime/src/link/primitive/join_link.rs:114]: Nah

File: route-rs-runtime/src/link/utils/task_park.rs
Comment (CollinValley) [reply route-rs-runtime/src/link/utils/task_park.rs]: not a bad idea. I'll try to think about some names that will work.

File: route-rs-runtime/src/utils/test/harness.rs
Comment (CollinValley) [reply route-rs-runtime/src/utils/test/harness.rs:49]: Oh, I can add the emoji back in, in later commits.

File: route-rs-runtime/src/utils/test/harness.rs
Comment (CollinValley) [reply route-rs-runtime/src/utils/test/harness.rs]: Yeah, in the later commit, I did just that, execute link was renamed to run_link in a couple commits.

File: route-rs-runtime/src/link/primitive/classify_link.rs
Comment (CollinValley) [reply route-rs-runtime/src/link/primitive/classify_link.rs:211]: Hmm, they are the same in the sense that you iterate over the entire collection of egressors, and then die_and_notify for each associated task park. So in this case, fork_link and classify_link have the same flow, and then join_link is its own case, and queue_link is the base case. So, I'm not sure there is much of a common piece we can cut out, without taking additional args into the function. 

I think I'm going to leave the the logic as is, while it isn't the most DRY, it is pretty readable right now.

File: route-rs-runtime/src/utils/test/harness.rs
Comment (CollinValley) [reply route-rs-runtime/src/utils/test/harness.rs:34]: Yeah, macros are kind of scary but yeah we should try to figure out some sort of macro. Will think about this for a future improvement.

File: examples/minimal-static-router/src/classifiers.rs
Comment (CollinValley) [reply examples/minimal-static-router/src/classifiers.rs:147]: Agree

File: route-rs-runtime/src/utils/test/harness.rs
Comment (CollinValley) [reply route-rs-runtime/src/utils/test/harness.rs:34]: Small issue is that the test fn's can not be async. Which is annoying, but we may be able to work around.

File: route-rs-runtime/src/link/utils/task_park.rs
Comment (CollinValley) [reply route-rs-runtime/src/link/utils/task_park.rs]: subbing wake should work pretty well, do we still like park though?

File: route-rs-runtime/src/link/primitive/join_link.rs
Comment (CollinValley) [reply route-rs-runtime/src/link/primitive/join_link.rs]: This ended up being closed by the harness rework. The harness could probably use some improvement via macro, but the initialize_runtime then runtime.block_on(... is likely what we are going to go with for the time being.

File: route-rs-runtime/src/link/mod.rs
Comment (ddowl) [reply route-rs-runtime/src/link/mod.rs:30]: > Structs wont implement unpin by default because they may contain pointers and such that could move. They could provide a custom implementation if that were the case, but since the ingressor and egressor own all their components, it isn't an issue to just add the trait to get the included 'provided' methods.

It feels like the compiler should be able to do some kind of analysis on structs to determine whether or not this is the case, but i don't wanna yell at the core team. I'm just happy `async` is stable.

That example was pretty informative. I'm going to watch withoutboats' talk on `Pin` to refresh my knowledge, but I'm going to trust your judgement 👌

File: route-rs-runtime/src/link/utils/task_park.rs
Comment (ddowl) [reply route-rs-runtime/src/link/utils/task_park.rs]: I still like "park".

File: route-rs-runtime/src/utils/test/harness.rs
Comment (ddowl) [reply route-rs-runtime/src/utils/test/harness.rs:49]: 💯

File: route-rs-runtime/src/utils/test/harness.rs
Comment (ddowl) [reply route-rs-runtime/src/utils/test/harness.rs:34]: That's so annoying! There will always be more "make testing better" work todo.

File: examples/dns-interceptor/src/pipeline.rs
Comment (ddowl) [reply examples/dns-interceptor/src/pipeline.rs]: :heck:

File: route-rs-runtime/src/utils/runner.rs
Comment (ddowl) [inline route-rs-runtime/src/utils/runner.rs]: >  for running the constructed router.TokioRunnable

typo?

File: route-rs-runtime/src/utils/runner.rs
Comment (ddowl) [inline route-rs-runtime/src/utils/runner.rs:50]: 🥇 yeet

File: route-rs-runtime/src/utils/runner.rs
Comment (CollinValley) [reply route-rs-runtime/src/utils/runner.rs]: yee, wil fixup

**Review by ddowl — DISMISSED** [review]
Looks like this branch needs another dose of `clippy` before Travis accepts it. Otherwise, LGGTMMD. Congrats on merging stable `async`!

**Comment by CollinValley** [comment]
I'll need to fixup graphgen tests before I am ready to fully merge, but it is close

File: route-rs-runtime/src/link/utils/task_park.rs
Comment (CollinValley) [reply route-rs-runtime/src/link/utils/task_park.rs]: named park_and_wake

File: examples/dns-interceptor/src/pipeline.rs
Comment (CollinValley) [reply examples/dns-interceptor/src/pipeline.rs]: gong to stick with simple for now here, when we revamp the new public facing running harness then I for one welcome our functional overlords.

**Comment by CollinValley** [comment]
This all looks good to me, will wait for Drew to do the honors

**Review by ddowl — APPROVED** [review]
gg Sam. Homie Stamp

**Pull request merged**
